### PR TITLE
Add toHTTPDate to DateTime class with supporting changes to CStringArray

### PR DESCRIPTION
**Pull request by riban-bw**
Corrects ISO8601 year to use 4 characters.
Add toHTTPDate function to DateTime class.
Calculate stringCount for CStringArray if not set, e.g. if object instantiated using assignment. Note: stringCount now mutable.
Add missing documentation for CStringArray class.
Use uppercase Y,M,D for date formats in documentation.

### CODE REVIEW

**Comment by riban-bw** [comment]
@slaff I avoided changing the coding style of the existing code (which was inconsistent and probably written before Sming project had agreed a style guide) so as to reduce the difficulty in identifying the functional changes to the library, i.e. ease validation and acceptance. Running
> make cs

results in lots of
> : CommandLine Error: Option 'disable-symbolication' registered more than once!
LLVM ERROR: inconsistency in registered CommandLine options

This may be because clang-tidy is not available for cygwin. I will try to tidy up the source code to match Sming coding style. I'll run it through clang-tidy on a Linux machine then update the PR.

File: Sming/Services/DateTime/DateTime.cpp [hunk 1]
```diff
@@ -148,7 +149,14 @@ String DateTime::toFullDateTimeString()
...
 {
 	char buf[32];
-	m_snprintf(buf, sizeof(buf), _F("%02d-%02d-%02dT%02d:%02d:%02dZ"), Year, Month + 1, Day, Hour, Minute, Second);
+	m_snprintf(buf, sizeof(buf), _F("%04d-%02d-%02dT%02d:%02d:%02dZ"), Year, Month + 1, Day, Hour, Minute, Second);
+	return String(buf);
+}
+
+String DateTime::toHTTPDate()
+{
+    char buf[30];
```
Comment (slaff) [inline Sming/Services/DateTime/DateTime.cpp]: Fix the indentation please. Should be as the line below.

File: Sming/SmingCore/Data/CStringArray.cpp [hunk 2]
```diff
@@ -1,55 +1,67 @@
...
+			if(index == 0)
+				return s;
+			offset += strlen(s) + 1;
+		}
+	}
+
+	return nullptr;
+}
+
+unsigned CStringArray::count() const
```
Comment (slaff) [inline Sming/SmingCore/Data/CStringArray.cpp]: Can you explain here why is that change needed?

File: Sming/Services/DateTime/DateTime.cpp [hunk 3]
```diff
@@ -4,7 +4,7 @@
```
Comment (slaff) [inline Sming/Services/DateTime/DateTime.cpp]: For consistency the following methods in the DateTime class should be renamed:

* parseHttpDate(const String& httpDate) -> fromHttpDate(const String& httpDate)
* convertFromUnixTime(time_t timep,..) -> fromFromUnixTime(time_t timep,..) 
* convertToUnixTime(int8_t sec, ..)  -> toUnixTime(int8_t sec, ..)

In addition if the DateTime class should deal with different date formatting then the following method could be a very good addition

`format(const String& format)`

Example:
```
DateTime d;
String httpDate = d.format("D, d M Y H:i:s GMT");
```

File: Sming/SmingCore/Data/CStringArray.h [hunk 4]
```diff
@@ -1,98 +1,122 @@
...
+		stringCount = 0;
+	}
+
+	/** @brief  Get quantity of strings in array
+	*   @retval unsigned Quantity of strings
+	*/
+	unsigned count() const;
+
+private:
+	mutable unsigned stringCount = 0;
```
Comment (slaff) [inline Sming/SmingCore/Data/CStringArray.h]: Why is this private variable set as mutable? Use-cases, please.

File: Sming/SmingCore/Data/CStringArray.h [hunk 5]
```diff
@@ -1,98 +1,122 @@
...
+	*   @retval bool True if string exists in array
+	*   @note   Search is not case-sensitive
+	*/
+	bool contains(const String& str) const
+	{
+		return indexOf(str) >= 0;
+	}
+
+	/** @brief Get string at the given position
+	 * 	@param index 0-based index of string to obtain
```
Comment (slaff) [inline Sming/SmingCore/Data/CStringArray.h]: Can you align `@param` with `@brief` in all places where you have added the documentation ?

File: Sming/SmingCore/Data/CStringArray.cpp (hunk 2 above)
Comment (riban-bw) [reply Sming/SmingCore/Data/CStringArray.cpp]: When an instance of CStringArray is created using an assignment, e.g. CStringArray cs("first\0second"); the quantity of elements is not set. This might normally be set within the constructor but I was unable to figure out how to do that in this case so instead I ensured count() returned the correct quantity of elements in the array by deducing whether it had already been calculated and if not then one-time calculating.

File: Sming/SmingCore/Data/CStringArray.h (hunk 4 above)
Comment (riban-bw) [reply Sming/SmingCore/Data/CStringArray.h]: count() is a const function so cannot change the value of the private variable stringCount unless stringCount is mutable. See explanation in other comment about why count() needs to set stringCount.

File: Sming/Services/DateTime/DateTime.cpp (hunk 1 above)
Comment (riban-bw) [reply Sming/Services/DateTime/DateTime.cpp]: Will do - IDE did its own thing. Will edit in text editor until I can whip IDE into line.

File: Sming/Services/DateTime/DateTime.cpp (hunk 3 above)
Comment (slaff) [reply Sming/Services/DateTime/DateTime.cpp]: > Do you want me to rationalise this?

Yes, but that should be done as a separate PR (after this one gets merged). The following refactoring changes should be done:
- Move DateTime from Services to SmingCore.
- Rename the old methods according to the rules above.
- In the DateTime.h header added the old methods that should redirect to the new ones, plus mark them as deprecated. Example:

```C++
class DateTime {
public: 
    /** @brief  Parse a HTTP full date and set time and date
     *  @param  httpDate HTTP full date in RFC 1123 format, e.g. Sun, 06 Nov 1994 08:49:37 GMT
     *  @retval bool True on success
     *  @note   Also supports obsolete RFC 850 date format, e.g. Sunday, 06-Nov-94 08:49:37 GMT where 2 digit year represents range 1970-2069
     *  @note   GMT suffix is optional and is always assumed / ignored
     *
     *   @deprecated Use `fromHttpDate` instead
     */
     bool parseHttpDate(const String& httpDate)   __attribute__ ((deprecated))
     {
            return fromHttpDate(httpDate);
     }
}
```

**Comment by slaff** [comment]
@mikee47 When you have a bit of time can you tell me what is your opinion about the suggested changes in CStringArray ?

File: Sming/SmingCore/Data/CStringArray.cpp [hunk 6]
```diff
@@ -1,55 +1,67 @@
...
+{
+	if(stringCount == 0 && length() > 0) {
+		//If array is created by assignment (e.g. CStringsArray csa = "Hello\0World";) then stringCount is not set so set it here
+		for(unsigned offset = 0; offset < length() + 1; offset++) {
```
Comment (mikee47) [inline Sming/SmingCore/Data/CStringArray.cpp]: why not just 'offset <= length()' ?

Comment (mikee47) [reply Sming/SmingCore/Data/CStringArray.cpp]: @slaff I like the changes, makes the class more useful.

Comment (slaff) [reply Sming/SmingCore/Data/CStringArray.cpp]: @riban-bw Can you make the suggested change ?

`for(unsigned offset = 0; offset < length() + 1; offset++) {`
to be
`for(unsigned offset = 0; offset <= length(); offset++) {`

Comment (riban-bw) [reply Sming/SmingCore/Data/CStringArray.cpp]: Doh! Done!

File: Sming/Services/DateTime/DateTime.h [hunk 7]
```diff
@@ -137,11 +137,20 @@ class DateTime
...
+     *  @retval String Date and time in format DD.MM.YYYY hh:mm:ss
      */
 	String toFullDateTimeString();
+
+   /** @brief  Get human readable date and time
+     *  @retval String Date and time in format YYYY-MM-DDThh:mm:ssZ
+     */
 	String toISO8601();
 
+   /** @brief  Get human readable date and time
```
Comment (slaff) [inline Sming/Services/DateTime/DateTime.h]: @riban-bw Open that file with very simple text editor and then find and fix all the places where the indentation is wrong. As above, for example.
